src/core/DY.Core.Trace.Base.fst

```diff
@@ -855,14 +814,79 @@ val trace_find_first: #label_t:Type -> tr:trace_ label_t -> e:trace_entry_ label
 let trace_find_first tr e =
   Some?.v (trace_search_first tr (trace_entry_equiv e))
 
-[@@"opaque_to_smt"]
-val trace_find_last: #label_t:Type -> tr:trace_ label_t -> e:trace_entry_ label_t{entry_exists tr e} ->
+val trace_find_last: #label_t:Type -> tr:trace_ label_t -> e:trace_entry_ label_t ->
     Pure timestamp
-      (requires True)
+      (requires equiv_entry_exists tr e)
       (ensures fun ts ->
         ts `on_trace` tr /\
         trace_entry_equiv e (get_entry_at tr ts) /\
         forall ts'. (ts' `on_trace` tr /\ trace_entry_equiv e (get_entry_at tr ts')) ==> ts' <= ts
       )
 let trace_find_last tr e =
   Some?.v (trace_search_last tr (trace_entry_equiv e))
+
+val opt_less_defined: #a:eqtype -> option a -> option a -> bool
+let opt_less_defined o1 o2 =
+  o1 = None || o1 = o2
+
+val trace_search_first_later:
+  #label_t:Type ->
+  tr1:trace_ label_t -> tr2:trace_ label_t ->
+  p:(trace_entry_ label_t -> bool) ->
+  Lemma
+  (requires tr1 <$ tr2)
+  (ensures opt_less_defined (trace_search_first tr1 p) (trace_search_first tr2 p))
+  [SMTPat (tr1 <$ tr2);
+   SMTPat (trace_search_first tr1 p);
+   SMTPat (trace_search_first tr2 p);
+  ]
+let trace_search_first_later tr1 tr2 p = ()
+
+val trace_find_first_later:
+  #label_t:Type ->
+  tr1:trace_ label_t -> tr2:trace_ label_t ->
+  e:trace_entry_ label_t ->
+  Lemma
+  (requires
+    equiv_entry_exists tr1 e /\
+    tr1 <$ tr2
+  )
+  (ensures trace_find_first tr1 e == trace_find_first tr2 e)
+  [SMTPat (tr1 <$ tr2);
+   SMTPat (trace_find_first tr1 e);
+   SMTPat (trace_find_first tr2 e);
+  ]
+let trace_find_first_later tr1 tr2 e = ()
+
```

Comment:
```suggestion
```

I would remove it since `trace_find_first_later`'s definition is (intentionally?) not opaque. In the rest of DY* we didn't write such `_later` lemmas for derived, non-opaque functions (such as `is_publishable`).

Reply:
But note that this is because this lemma is expected to only be instantiated via SMT pattern, for lemmas that are called manually (such as others in `Arithmetic.Derived` I think it makes sense to keep both versions.